Observable Item Edit and View

nspool opened the pull request:
Implements the remaining Observable Item screens.
Add a new SearchInput component that SupersededBy with the NRMN search endpoint to autocomplete

utas-tajona commented on api/src/main/java/au/org/aodn/nrmn/restapi/controller/ObservableItemController.java:
```diff
@@ -50,9 +53,23 @@ public ObservableItemDto newObservableItem(@Valid @RequestBody ObservableItemDto
+
+        if(sitePostDto.getSpeciesEpithet() == null || sitePostDto.getSpeciesEpithet().isEmpty())
+            errors.add(new ValidationError(ObservableItemDto.class.getName(), "speciesEpithet", null, "Species Epithet is required."));
+        
```
Validated is here to do that in more gracefull way

jonescc replied:
Actually this should just be a javax validation - there's no need to complicate this with validated/turning validated into a response (yes I said complicate)

nspool replied:
Will use javax validation.

jonescc commented on api/src/main/java/au/org/aodn/nrmn/restapi/controller/ObservableItemController.java:
```diff
@@ -77,4 +93,21 @@ private void validatePost(@RequestBody @Valid ObservableItemDto sitePostDto) {
+
+        if(!errors.isEmpty())
+            throw new ValidationException(errors);
+    }
```
This is actually doing nothing - I'd remove it.

nspool replied:
Can you please provide an example of how this is done using Validated?

nspool replied:
Refactored so the same validation is used on PUT and POST. Empty method removed.

nspool commented on api/src/main/java/au/org/aodn/nrmn/restapi/controller/ObservableItemController.java:
```diff
@@ -50,9 +53,23 @@ public ObservableItemDto newObservableItem(@Valid @RequestBody ObservableItemDto
+    
+    @GetMapping("/observableItem/{id}")
+    public ObservableItemGetDto findOne(@PathVariable Integer id) {
+        ObservableItem observableItem = observableItemRepository.findById(id).orElseThrow(ResourceNotFoundException::new);
```
So just remove the `.orElseThrow(ResourceNotFoundException::new);` ?

utas-tajona replied:
```
return observableItemRepository.findById(id).map(item -> ResponseEntity.ok.body(pper.map(item, ObservableItemGetDto,class)).orElseGet( () -> ResponseEntity.notFound().build() ;
```

should work

nspool replied:
Thanks, Fixed.

nspool commented:
Left validation as-is for now as I couldn't work out a way to do the validation with Cyclops that was any more concise.
Moved POST validations to javax and fixed missing uniqueness validations on PUT.

nspool commented:
http://nrmn-obs-item-view-edit.dev.aodn.org.au/

jonescc commented:
Has merge conflicts

nspool commented:
@jonescc Fixed.

jonescc commented on api/src/main/java/au/org/aodn/nrmn/restapi/controller/mapping/ObservableItemGetDtoMapperConfig.java:
```diff
@@ -0,0 +1,43 @@
+    @Autowired
+    ObservableItemGetDtoMapperConfig(ObsItemTypeRepository obsItemTypeRepository, ObservableItemRepository observableItemRepository, ModelMapper modelMapper) { 
+        
+        Converter<ObsItemType, String> toObsItemTypeName = ctx -> ctx.getSource() == null ? null : ctx.getSource().getObsItemTypeName();
```
Shouldn't be required if dto properties are named so that object mapper can map by default.  You can always use a json alias to rename to make the response nicer

jonescc commented on api/src/main/java/au/org/aodn/nrmn/restapi/controller/mapping/ObservableItemGetDtoMapperConfig.java:
```diff
@@ -0,0 +1,43 @@
+        Converter<ObsItemType, Integer> toObsItemTypeId = ctx -> ctx.getSource() == null ? null : ctx.getSource().getObsItemTypeId();
+        Converter<AphiaRelType, String> toAphiaRelTypeName = ctx -> ctx.getSource() == null ? null : ctx.getSource().getAphiaRelTypeName();
+        Converter<Integer, String> toSupersededByIDs = ctx -> observableItemRepository.findOneProjectedById(ctx.getSource()).getSupersededIDs();
+        Converter<Integer, String> toSupersededByNames = ctx -> observableItemRepository.findOneProjectedById(ctx.getSource()).getSupersededNames();
```
I'd use a postConverter so that we only perform the lookup once.   See the species mapper.

utas-tajona reviewed (approved):
lgtm
